### PR TITLE
add object support for 'and' and 'or' functions. Closes #344

### LINKED ISSUE #344: [feature request] support object as parameter in where()
is it possible to filter line by object? like
```ts
interface db = {
  test: TestTable
}
interface TestTable = {
  filterParam: number
}
kysely.selectFrom('test')
  .where({ filterParam: 1 })
  .selectAll()
  .execute()
```

### PR DESCRIPTION
**Pull request by koskimas**
Support

```ts
db.selectFrom('person')
  .selectAll()
  .where((eb) => eb.and({
    first_name: 'Jennifer',
    last_name: eb.ref('first_name')
  }))
```

Same works with `eb.or`.

### CODE REVIEW

File: src/expression/expression-builder.ts [hunk 1]
```diff
@@ -812,45 +866,23 @@ export function createExpressionBuilder<DB, TB extends keyof DB>(
...
-        exprs[1].toOperationNode()
-      )
-
-      for (let i = 2; i < exprs.length; ++i) {
-        node = AndNode.create(node, exprs[i].toOperationNode())
+      if (isReadonlyArray(exprs)) {
+        return new ExpressionWrapper(parseFilterList(exprs, 'and'))
+      } else {
+        return new ExpressionWrapper(parseFilterObject(exprs, 'and'))
       }
```
Comment (igalklebanov) [inline src/expression/expression-builder.ts:875]:
```suggestion

      return new ExpressionWrapper(parseFilterObject(exprs, 'and'))
```

File: src/expression/expression-builder.ts [hunk 2]
```diff
@@ -812,45 +866,23 @@ export function createExpressionBuilder<DB, TB extends keyof DB>(
...
-        exprs[1].toOperationNode()
-      )
-
-      for (let i = 2; i < exprs.length; ++i) {
-        node = OrNode.create(node, exprs[i].toOperationNode())
+      if (isReadonlyArray(exprs)) {
+        return new ExpressionWrapper(parseFilterList(exprs, 'or'))
+      } else {
+        return new ExpressionWrapper(parseFilterObject(exprs, 'or'))
       }
```
Comment (igalklebanov) [inline src/expression/expression-builder.ts:885]:
```suggestion

      return new ExpressionWrapper(parseFilterObject(exprs, 'or'))
```

File: src/parser/binary-operation-parser.ts [hunk 3]
```diff
@@ -94,13 +112,50 @@ export function parseReferentialBinaryOperation(
...
+export function parseFilterObject(
+  obj: Readonly<FilterObject<any, any>>,
+  combinator: 'and' | 'or'
+): OperationNode {
+  return parseFilterList(
+    Object.entries(obj)
+      .filter(([, v]) => !isUndefined(v))
+      .map(([k, v]) =>
+        parseValueBinaryOperation(k, needsIsOperator(v) ? 'is' : '=', v)
+      ),
```
Comment (igalklebanov) [inline src/parser/binary-operation-parser.ts:124]: nitpicky

```suggestion
    Object.entries(obj)
      .reduce((parsed, [k, v]) => {
        if (isUndefined(v)) return acc
        
        const op = needsIsOperator(v) ? 'is' : '='
        
        parsed.push(parseValueBinaryOperation(k, op, v))
        return parsed
      }, []),
```

**Review by igalklebanov — APPROVED** [review]
Great improvement to DX! 🎉 

LFGTM! 🚀 

left some suggestions..

File: src/parser/binary-operation-parser.ts (hunk 3 above)
Comment (koskimas) [reply src/parser/binary-operation-parser.ts:124]: Hmm... I think the current one is cleaner. Isn't the `parsed` array also `any[]` here?

Comment (igalklebanov) [reply src/parser/binary-operation-parser.ts:124]: Yeah, TypeScript doesn't play nice with reduce, you usually have to `as T` the initial value.
This was just a micro-optimization (single pass vs. two passes).

**Pull request merged**